File: stack/394.decode-string.py
#
# @lc app=leetcode id=394 lang=python3
#
# [394] Decode String
#

# @lc code=start


class Solution:
    def decodeString(self, s: str) -> str:
        numStack = []
        strStack = []
        tail = ""
        i = 0
        while i < len(s):
            if s[i].isdigit():
                n = int(s[i])
                while s[i + 1].isdigit():
                    n = 10 * n + int(s[i + 1])
                    i += 1
                numStack.append(n)
            elif s[i] == "[":
                strStack.append(tail)
                tail = ""
            elif s[i] == "]":
                tmp = strStack.pop()
                times = numStack.pop()
                tmp = tmp + tail * times
                # strStack.append(tmp)
                tail = tmp
            else:
                tail += s[i]
            i += 1
        return tail


# @lc code=end
if __name__ == "__main__":
    s = Solution()
    print(s.decodeString("3[a]2[bc]"))

File: stack/394.decode-string.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

/*
 * @lc app=leetcode id=394 lang=golang
 *
 * [394] Decode String
 */

// @lc code=start
func decodeString(s string) string {
	var numStack []int
	var strStack []string
	tail := ""
	for i := 0; i < len(s); i++ {
		switch {
		case isDigit(string(s[i])):
			n, _ := strconv.Atoi(string(s[i]))
			for isDigit(string(s[i+1])) {
				n = 10*n + int(s[i+1])
				i++
			}
			numStack = append(numStack, n)
		case s[i] == '[':
			strStack = append(strStack, tail)
			tail = ""
		case s[i] == ']':
			tmp := strStack[len(strStack)-1]
			strStack = strStack[:len(strStack)-1]
			times := numStack[len(numStack)-1]
			numStack = numStack[:len(numStack)-1]
			tmp += strings.Repeat(tail, times)
			tail = tmp
		default:
			tail += string(s[i])
		}
	}
	fmt.Println(strStack)
	fmt.Println(numStack)
	return tail
}

func isDigit(s string) bool {
	if _, err := strconv.Atoi(s); err != nil {
		return false
	}
	return true
}

func main() {
	fmt.Println(decodeString("3[a]2[bc]"))
}

// @lc code=end
